### PR TITLE
Bug Fix: Possibly Undefined Checker

### PR DESCRIPTION
**Pull request by kavins14**
- Fixes issue #657 

Consider the following python program:
```python 
[x for x in lst] 
```
- It turns out that that the algorithm was taking into account the assignment of `x` from `for x in lst` AFTER taking into account the expression `x` making it seem as though `x` was not defined before being used. This is because the control flow graph does not take into consideration the order in which the sub-statements are evaluated. 
- A similar issue arises in the following statement `x = x + 1`. The algorithm would first consider `x` being assigned then `x + 1` being evaluated. It should actually be the other way around.

This PR fixes both these issues.

### CODE REVIEW

**Review by david-yz-liu — COMMENTED** [review]
Hi @kavins14! Thanks for looking into this. It seems to me that you've correctly identified the (two!) problems with the current checker, but I think the solution could be cleaner. Right now the `get_nodes` function is really relying on the fact that it's only being called on simple statements to ensure the overall ordering of `Name`s/`AssignName`s is correct. (For example, if `statement` was a for loop containing multiple assignment statements, you can't extract all of the `Name` nodes and yield them first.)

Given this, I think it's better to be more explicit about how specific node classes require a different ordering than "nodes of class" provides. I would imagine something like:

```python
def get_nodes(self, statement: astroid.node_classes.NodeNG) -> Generator[astroid.node_classes.NodeNG, None, None]:
    if isinstance(statement, astroid.Assign):
        # Yield from calling nodes_of_class on RHS, then yield target(s)
    elif isinstance(statement, (astroid.ListComprehension, astroid.ListComprehension, ...):
        # Yield from calling nodes_of_class on comprehensions, then yield body
    else:
        # Yield from calling nodes_of_class on statement
```

It's less elegant to have to list out the comprehension expression types, but overall I think the logic will be more clear this way.

**Comment by kavins14** [comment]
You are absolutely right @david-yz-liu, implementing the function that way seems to be the right way to do it.

File: python_ta/checkers/possibly_undefined_checker.py [hunk 1]
```diff
@@ -119,6 +118,19 @@ def _get_assigns(self, node: Union[astroid.FunctionDef, astroid.Module]) -> Set[
 
         return assigns.difference(kills)
 
+    def get_nodes(self, statement: astroid.node_classes.NodeNG) -> Generator[astroid.node_classes.NodeNG, None, None]:
+        multiple_nodes = lambda nodes : chain.from_iterable(self.get_nodes(node) for node in nodes)
+        if isinstance(statement, astroid.Assign):
+            # RHS is evaluated before assigned in an assignment statement
+            yield from self.get_nodes(statement.value)
+            yield from multiple_nodes(statement.targets) # statement.targets is a list of nodes
+        elif isinstance(statement, (astroid.ListComp, astroid.SetComp)):
```
Comment (david-yz-liu) [inline python_ta/checkers/possibly_undefined_checker.py:127]: Please add `astroid.DictComp` and `astroid.GeneratorExp` here

File: tests/test_custom_checkers/test_possibly_undefined_checker.py [hunk 2]
```diff
@@ -393,3 +393,42 @@ def func2():
...
+        print(list(func_node.nodes_of_class(astroid.Name)))
+
+        self.checker.visit_functiondef(func_node)
+        with self.assertAddsMessages(
+            pylint.testutils.Message(
+                msg_id='possibly-undefined',
+                node=name_node_x,
+            )
+        ):
+            self.checker.visit_name(name_node_x)
```
Comment (david-yz-liu) [inline tests/test_custom_checkers/test_possibly_undefined_checker.py:434]: Please insert a blank line at the end of this file.

**Review by david-yz-liu — COMMENTED** [review]
Thanks @kavins14! I left a few minor comments, but once you're done that I will merge this in!

**Pull request merged**